lib/rucio/client/commands/rse.py

```diff
@@ -0,0 +1,283 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import json
+from typing import TYPE_CHECKING
+
+from rucio.client.commands.bin_legacy.rucio import list_rses
+from rucio.client.commands.bin_legacy.rucio_admin import (
+    add_distance_rses,
+    add_protocol_rse,
+    add_qos_policy,
+    add_rse,
+    del_protocol_rse,
+    delete_attribute_rse,
+    delete_distance_rses,
+    delete_limit_rse,
+    delete_qos_policy,
+    disable_rse,
+    get_attribute_rse,
+    get_distance_rses,
+    info_rse,
+    list_qos_policies,
+    set_attribute_rse,
+    set_limit_rse,
+    update_distance_rses,
+    update_rse,
+)
+from rucio.client.commands.command_base import CommandBase
+
+if TYPE_CHECKING:
+    from argparse import ArgumentParser
+
+    from rucio.client.commands.utils import OperationDict
+
+
+class RSE(CommandBase):
+    def module_help(self) -> str:
+        return "Manage Rucio Storage Elements - the sites where Rucio can place and access data."
+
+    def _operations(self) -> dict[str, "OperationDict"]:
+        return {
+            "list": {"call": self.list_, "docs": "Show all RSEs.", "namespace": self.list_namespace},
+            "show": {"call": self.show, "docs": "Show history of RSE usage (date created, date updated, etc)", "namespace": self.namespace},
+            "remove": {"call": self.remove, "docs": "Disable an RSE.", "namespace": self.namespace},
+            "add": {"call": self.add, "docs": "Create a new RSE.", "namespace": self.namespace},
+            "update": {"call": self.update, "docs": "Update an existing RSE.", "namespace": self.namespace},
+        }
+
+    def implemented_subcommands(self) -> dict[str, type[CommandBase]]:
+        return {"attribute": Attribute, "distance": Distance, "protocol": Protocol, "limit": Limit, "qos": QOS}
+
+    def usage_example(self) -> list[str]:
+        return [
+            "$ rucio rse  # Show all current RSEs, can also access with `rucio rse list`",
+            "$ rucio rse list --rse 'deterministic=True' # Show all RSEs that match the RSE Expression 'deterministic=True'",
+            "$ rucio rse remove --rse RemoveThisRSE  # Disable an RSE by name",
+            "$ rucio rse add --rse CreateANewRSE  # add a new RSE named CreateANewRSE",
+            "$ rucio rse update --rse rse123456 --setting deterministic --value False  # Make an RSE Non-Deterministic",
+            "$ rucio rse show --rse rse123456  # See all the settings for rse123456"
+        ]
```

Comment:
I also think this is confusing. There is a number of places where both RSE and RSE Expressions are allowed. How best do we handle this? This confusion here is symptomatic of that.

Reply:
> both RSE and RSE Expressions

An RSE is a unitary RSE expression. So the question becomes not when both are allowed, but only the former (i.e. the argument is an RSE name and not an RSE expression).

This was originally reported in #1839. The solution was to use `--rse` when a single RSE name is expected and `--rses` when it is an RSE expression.

Reply:
I'm afraid I'm very unsatisfied with that solution. There's no reason that `rses` would imply it's anything other than a list of rse names. Perhaps the more restrictive version (e.g. only allows a single explicit rse name) should just be `--rse-name` and the generic rse expression or explicit rse name should be `--rse-expression` in all cases?

Reply:
Have to agree with @voetberg here. One of my biggest gripes with the current CLI is the inconsistency of `--rse`, `--rses` and `--rse-expression`.

There's two cases: one where we definitely only want to interact with a single RSE, in that case positional makes sense IMHO. (I could be convinced for `--rse` if you really want.) For everything else this is always a set based operation, so here we need to merge `--rses` and `--rse-expression` into a single mindful variant. Since this also includes RSE attributes, I would rather tend towards `--rse-expression` but I find that a bit clunky. How about `--rse-filter`?

Reply:
I would have gone with `--rse-expression` as well, but that’s what was decided the last time. I think the idea was that, wherever you can put multiple RSEs, you do so exclusively via an RSE expression.

So the problem here is that you can either follow that decision or propose something that is universally-accepted as superior. And the argument against `--rse-expression` is its length.

I have to say, I find the documentation provided in the original `rucio list-dataset-replicas` on-point. Hence, I’ve grown accustomed to `--rses`.

Reply:
> One of my biggest gripes with the current CLI is the inconsistency of --rse, --rses and --rse-expression.

The current CLI *is* consistent. There’s no `--rse-expression` anymore.

Reply:
ha! just shows how often i've been using it!

so... `--rses` for the set-based selection, and for single RSEs positional and/or `--rse` ?

Reply:
Yes, that is correct. If there is some command that behaves differently, then that would qualify as a bug.

Reply:
> > One of my biggest gripes with the current CLI is the inconsistency of --rse, --rses and --rse-expression.
> 
> The current CLI _is_ consistent. There’s no `--rse-expression` anymore.

There are instances of --expression or `rse_expression` as a positional arg (`list-suspicious-replicas`, `add-replication-rule`).

I still think `rse` and `rses` is not sufficient in explaining the differences between the two, so many pluralized nouns are just stylistic or inconsistent (in may code bases, not just ours), it's strange to make people assume that this is _the_ instance where the `s` matters. But if this is the solutions those brave programmers of 1839 decided, I will defer to that decision.

Reply:
I should note, you can have multiple kwargs matching to a variable. We have it in `declare-temporary-unavialable-replica`, albeit for a depreciating argument. 
```
unavailable_parser.add_argument("--expiration-date", "--duration", new_option_string="--duration", dest="duration", action=StoreAndDeprecateWarningAction, type=int, help="Timeout in seconds when the replicas will become available again")

```

Reply:
> There are instances of `--expression` or `rse_expression` as a positional arg (`list-suspicious-replicas`, `add-replication-rule`).

I’m afraid I stand by my comment. `list-suspicious-replicas` was introduced in parallel with the effort to unify the interface, hence it was missed. But I’m willing to bet my life savings that nobody is using this command. In `add-replication-rule` it is the description of an argument that matches the technical term. The only area where the interface is not consistent is some `rucio-admin` commands that offer it either as a keyword argument (e.g. `rucio-admin rse set-attribute` or as a positional argument (e.g. `rucio-admin rse add-protocol`).

I’m not deaf to your arguments; I would have chosen `--rse-expression` myself. But that was the proposal which was approved and accepted (recently, I might add) and the commands for the daemons follow it. Hence, we’re dealing with 1) a proposal that is not unequivocally superior to the current implementation and 2) change aversion.